pkg/tagset/caching_factory.go

```diff
@@ -0,0 +1,176 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.Datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package tagset
+
+import (
+	"strings"
+
+	"github.com/twmb/murmur3"
+)
+
+// A cachingFactory is a Factory implementation that caches Tags instances.
+// See NewCachingFactory for usage.
+//
+type cachingFactory struct {
+	baseFactory
+
+	// Tags instances are cached by 64-bit cache keys that can have a range of
+	// meanings; each CacheID identifies a different such meaning.  Caches with
+	// different CacheIDs are stored independently.
+	caches [numCacheIDs]tagsCache
+}
+
+// NewCachingFactory creates a new caching factory.  A caching factory caches
+// Tags instances when they are seen, and uses those cached values when possible
+// to reduce CPU and memory usage.
+//
+// The size of the factory's cache is proportional to cacheSize: double this
+// parameter and memory usage will roughly double.  The cache size must be at
+// least 1.
+//
+// For a given cache size, the cacheWidth parameter determines how well the
+// cache handles eviction.  If set to 1, the entire cache is thrown out when it
+// is full.  Larger values do a better job of holding on to
+// less-frequently-referenced values, at the expense of more CPU time spent
+// searching for cached instances.  A value of 5 is a nice starting point.  The
+// cache width must be at least 1.
+func NewCachingFactory(cacheSize, cacheWidth int) Factory {
+	if cacheSize < 1 {
+		panic("cacheSize must be at least 1")
+	}
+	if cacheWidth < 1 {
+		panic("cacheWidth must be at least 1")
+	}
```

Comment:
Why not return an error instead? I think it warrants a comment

Reply:
These are programming errors, rather than runtime errors -- in other words, these are invariants.  I added them only because (at least in one iteration of this code) they would otherwise cause an unterminated loop.  I could remove them, but I don't think this method should return an error if that error can practically never occur.

Reply:
I admit I just find it scary to have panics on a function in the public API of the module, but it's true that this values will be hardcoded in the program in any case I can think of, so we can leave it like this

Reply:
Yeah, I understand the feeling!  It'd be nice if there was a way to do this only in dev builds, like rust's `debug_assert!(..)`.

The minima are also in the docs.  I think there are cases like this in the stdlib, too -- "must not be copied after initialized", "Reset must be called before use", etc.  Some of those lead to UB (which could include a panic), and some explicitly panic.

Reply:
+1. `newTagsCache` is already returning errors in a similar way so it would be consistent. Plus since this functiona already returns an `error` (that the caller has to handle) it seems pretty low impact to move these panics to errors.